Adding nand flash support, based on dhara nand library (IDFGH-7900)

mikkeldamsgaard opened the pull request:
Adding basic nand support and enable mounting of VFS FAT on top of nand.

Uses dhara library

The only hw available to us at build time was the winbond chip. The rest has been added based purely on datasheet information and has not been verified

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/espressif/esp-idf?pullRequest=9422) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/espressif/esp-idf?pullRequest=9422) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/esp-idf?pullRequest=9422) it.</sub>

igrr commented:
Hi @mikkeldamsgaard,

Thanks a lot for the pull request! It looks really useful, support for NAND flash has been on our backlog for a long time, so this is a welcome feature.

I have a few notes about the logistics of getting such feature supported in ESP-IDF:

- (trivial) we accept PRs trageted at `master` branch only. For bugfix PRs we can also backport the changes to an older release branch, on request. For this reason, we can't accept this feature on release/v4.4.
- An alternative is to make this feature available as a component outside of ESP-IDF. In IDF v5.0 (to be released soon) we are moving multiple IDF components out of the source tree; they will be developed and maintenaned in separate repositories. This allows us to decouple the release cycles of these component from the release cycle of IDF. This also allows users of older IDF version to benefit from some of the new components! One of the repositories where such external components are maintained is https://github.com/espressif/idf-extra-components. We can try to merge this new component there. The component can then be added to the project using the IDF components manager (https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/tools/idf-component-manager.html, https://components.espressif.com/)
- We will need to get the hardware to test the NAND flash driver on our end, and find developers internally who would maintain the driver in the future. This might take some time.
- We also need to do a proper review of the newly added code, to be sure we can understand it well and can support customers who might end up using it, in case they run into any issue.

Sorry if this all sounds complex! We'll try to get the changes properly reviewed and merged. We'll keep you updated about the progress.

mikkeldamsgaard commented:
It is unclear to me if you expect me to take any actions for now.

pacucha42 commented:
@mikkeldamsgaard - there's no action needed on your side for now, we are preparing necessary HW setup and will test your PR shortly. We will let you know as soon as possible, see @igrr 's previous comment. Please, hold on.
Thank you very much for your effort!

chaabanemalki commented:
This is very interesting indeed. I was thinking to write a driver for my Winbond chip W25N01GVZEIG.

Thank you a lot @mikkeldamsgaard  .

I'm wondering what is the best way to port the changes to my esp-idf repository coming from official repos ?

Any recommandations ?

mikkeldamsgaard commented:
It is on my roadmap to upgrade this to idf-master. It is unclear what the timeline is for me.
If you want to look at it, it should not be that terribly difficult, as it is mostly new files. The changes to existing files are pretty minimal. It does use some other idf functions, so if they changed a lot, there would be some work involved.

newbloodmage commented on components/spi_nand_flash/src/spi_nand_oper.c:
```diff
@@ -0,0 +1,125 @@
+  return spi_nand_execute_transaction(device, &t);
+}
+
+esp_err_t spi_nand_erase_block(spi_device_handle_t device, uint16_t page) {
```
For W32N02KV etc, page size exceeds range of uint16_t

cennian replied:
For the W25N02KV I only get half the flash size on the FAT FS. How can that be resolved?

pacucha42 replied:
Thanks for pointing this out. It sounds like a mistake, spi_nand_transaction_t.address is uint32_t. Let us check whole context, please, we shall update you asap. This PR is currently being processed and not yet published as IDF part

RathiSonika commented:
Hello @mikkeldamsgaard, 
Thank you so much for your pull request. We've been actively working on it. I have rebased it to the latest master with some minor changes and generated a patch. Kindly review the changes provided in the patch [0001-feat-spi_nand_flash-Adding-nand-flash-support-based-.patch](https://github.com/espressif/esp-idf/files/13727392/0001-feat-spi_nand_flash-Adding-nand-flash-support-based-.patch) and share any comments you may have. Following your review, our plan is to incorporate it as a component in the [IDF component registry](https://components.espressif.com/). 
Additionally, the review comment mentioned above is valid and I have addressed it in the same patch. Kindly verify the changes.
Thank you!

cennian commented:
> Hello @mikkeldamsgaard, Thank you so much for your pull request. We've been actively working on it. I have rebased it to the latest master with some minor changes and generated a patch. Kindly review the changes provided in the patch [0001-feat-spi_nand_flash-Adding-nand-flash-support-based-.patch](https://github.com/espressif/esp-idf/files/13727392/0001-feat-spi_nand_flash-Adding-nand-flash-support-based-.patch) and share any comments you may have. Following your review, our plan is to incorporate it as a component in the [IDF component registry](https://components.espressif.com/). Additionally, the review comment mentioned above is valid and I have addressed it in the same patch. Kindly verify the changes. Thank you!

Has the demo been created? The link is non-existent. You haven't addressed the issue in the patch that @newbloodmage mentioned

RathiSonika commented:
Hello @cennian,
The demo is not included in the patch; it primarily consists of the rebased code with the latest ESP-IDF master and minor improvements. I am seeking a review and confirmation from the author before proceeding and adding it to the component registry, attributing the changes to the author's name. We are in the process of working on the demo.

Concerning the review comment with `esp_err_t spi_nand_erase_block(spi_device_handle_t device, uint16_t page)`, this function should have the data type of page updated to uint32_t, which is addressed in the patch.

mikkeldamsgaard commented:
> Hello @mikkeldamsgaard, Thank you so much for your pull request. We've been actively working on it. I have rebased it to the latest master with some minor changes and generated a patch. Kindly review the changes provided in the patch [0001-feat-spi_nand_flash-Adding-nand-flash-support-based-.patch](https://github.com/espressif/esp-idf/files/13727392/0001-feat-spi_nand_flash-Adding-nand-flash-support-based-.patch) and share any comments you may have. Following your review, our plan is to incorporate it as a component in the [IDF component registry](https://components.espressif.com/). Additionally, the review comment mentioned above is valid and I have addressed it in the same patch. Kindly verify the changes. Thank you!

Thank you for picking this up.

After I submitted the original PR I had a chance to test the code even more and we found an issue in multi-threaded scenarios. Some of the methods in nand.c needs to have a mutex protecting them. I added the relevant code here: https://github.com/mikkeldamsgaard/esp-idf/blob/spottune/components/spi_nand_flash/src/nand.c

Other than that, I do not have any comments.

pacucha42 commented:
> After I submitted the original PR I had a chance to test the code even more and we found an issue in multi-threaded scenarios. Some of the methods in nand.c needs to have a mutex protecting them. I added the relevant code here: https://github.com/mikkeldamsgaard/esp-idf/blob/spottune/components/spi_nand_flash/src/nand.c
> 
> Other than that, I do not have any comments.

Thank you very much for the info, we shall update the source file accordingly

RathiSonika commented:
@mikkeldamsgaard Please take a look to this [PR- feat(nand_flash): Add nand flash support, based on dhara nand library ](https://github.com/espressif/idf-extra-components/pull/290) and can you please sign the CLA [here](https://github.com/espressif/idf-extra-components/pull/290#issuecomment-1875238092) for the same?

chaabanemalki commented:
Can someone clarify the difference between this PR and https://github.com/espressif/idf-extra-components/pull/290 from @RathiSonika 
This PR is for esp-idf master branch but the other one is for [idf-extra-components](https://github.com/espressif/idf-extra-components) repository that I'm not sure how to use ?

RathiSonika commented:
@chaabanemalki The SPI NAND Flash component will be included in the component registry. You can find detailed information about it at this link: [ESP-IDF Component Manager and ESP Component Registry Documentation](https://docs.espressif.com/projects/idf-component-manager/en/latest/#idf-component-manager-and-esp-component-registry-documentation).

chaabanemalki commented:
> @chaabanemalki The SPI NAND Flash component will be included in the component registry. You can find detailed information about it at this link: [ESP-IDF Component Manager and ESP Component Registry Documentation](https://docs.espressif.com/projects/idf-component-manager/en/latest/#idf-component-manager-and-esp-component-registry-documentation).

Very helpful. thank you

RathiSonika commented:
This pull request has been merged and is now included in the commit [339b31c](https://github.com/espressif/idf-extra-components/commit/339b31cc0d653b898b6fd9c66de27186dfc46366). You can find this component listed in the [component registry](https://components.espressif.com/components/espressif/spi_nand_flash).
